### PR TITLE
AC-5624 - Allow Criterion objects to be edited and copied in the Application Allocation frontend

### PR DESCRIPTION
**Pull request by jonkiparsky**
To test: with impact api and application allocator running on development machine, load
http://localhost:1234/set-criteria

Note: If [AC-5906  on impact-api](https://github.com/masschallenge/impact-api/pull/194) has not merged, you will need to be on that branch for cloning to work. 

* Should see no criteria, as judging round 56 is hard-coded, and no criteria exist for this round on the test db. (when this is connected to app-allocator-setup, judging round will be passed in as a prop)
* in a `make django-shell` under impact-api, should see
```
In [5]: mc.CriterionOptionSpec.objects.values_list("option", "criterion__judging_round_id")
Out[5]: <QuerySet [('', 48), ('', 48), ('', 48), ('Executive', 48), ('Investor', 48), ('Lawyer', 48), ('female', 48), ('male', 48)]>
```
(note I am importing accelerator.models as mc, from force of habit)
* select "2017-04 BOS CH IL" round
* hit "clone"
* after a moment, page will get the new criteria and you'll see something like the attached screen shot
* In django shell, should now see
```
In [7]: mc.CriterionOptionSpec.objects.values_list("option", "criterion__judging_round_id")
Out[7]: <QuerySet [('', 48), ('', 48), ('', 48), ('Executive', 48), ('Investor', 48), ('Lawyer', 48), ('female', 48), ('male', 48), ('', 56), ('', 56), ('', 56), ('Executive', 56), ('Investor', 56), ('Lawyer', 56), ('female', 56), ('male', 56)]>
```
* You should now be able to edit weight and count of criteria. On hitting the submit button, you will be redirected to the app-allocator-setup screen, and your changes will be visible there. (note that the criteria load slowly)
* going to /set-criteria again will show your edits have persisted there as well. 

screen shot 2018-08-15 at 1 47 27 pm

### CODE REVIEW

**Comment by pbx** [comment]
Note: I'm seeing behavior from the client that looks like polling, which I take it is not intended:
```[15/Aug/2018 16:09:31] "GET /api/v1/analyze_judging_round/56/ HTTP/1.0" 200 14
[15/Aug/2018 16:09:32] "GET /api/v1/analyze_judging_round/56/ HTTP/1.0" 200 14
[15/Aug/2018 16:09:32] "GET /api/v1/analyze_judging_round/56/ HTTP/1.0" 200 14
[15/Aug/2018 16:09:32] "GET /api/v1/analyze_judging_round/56/ HTTP/1.0" 200 14
[15/Aug/2018 16:09:32] "GET /api/v1/analyze_judging_round/56/ HTTP/1.0" 200 14
[15/Aug/2018 16:09:32] "GET /api/v1/analyze_judging_round/56/ HTTP/1.0" 200 14
[15/Aug/2018 16:09:32] "GET /api/v1/analyze_judging_round/56/ HTTP/1.0" 200 14
[15/Aug/2018 16:09:33] "GET /api/v1/analyze_judging_round/56/ HTTP/1.0" 200 14
[15/Aug/2018 16:09:33] "GET /api/v1/analyze_judging_round/56/ HTTP/1.0" 200 14
[15/Aug/2018 16:09:33] "GET /api/v1/analyze_judging_round/56/ HTTP/1.0" 200 14
```
(This is impact-api runserver console output, with the http://localhost:1234/set-criteria page loaded)

File: src/SetCriteriaHeader.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import React from 'react';
+
+
+class SetCriteriaHeader extends React.Component {
+    render () {
+	return (
+		<div>
+		<h1>Set Judging Round Criteria </h1>
```
Comment (shanbady) [inline src/SetCriteriaHeader.js:8]: this doesn't functionally do anything other than render static html. is it necessary to break out into a component?

File: src/SetCriteria.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        source_judging_round_id: this.state.clone_judging_round,
+        target_judging_round_id: this.judging_round
+      })
+    }).then(this.setState({ key: Math.random() }));
+  };
+
+  render() {
+    return (
+      <div>
+            <SetCriteriaHeader />
```
Comment (shanbady) [inline src/SetCriteria.js:39]: formatting.

File: src/SetCriteria.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+      mode: "cors",
+      method: "post",
+      headers: {
+        "Content-Type": "application/json; charset=utf-8"
+      },
+      body: JSON.stringify({
+        source_judging_round_id: this.state.clone_judging_round,
+        target_judging_round_id: this.judging_round
+      })
+    }).then(this.setState({ key: Math.random() }));
```
Comment (shanbady) [inline src/SetCriteria.js:33]: why is this necessary?

File: src/SetCriteria.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+class SetCriteria extends React.Component {
+  constructor(props) {
+    super(props);
+    this.judging_round = 56; // change this once app-allocator-setup is calling this page
+    this.cloneCriteria = this.cloneCriteria.bind(this);
+  }
+  state = {
+    judging_round: null
+  };
```
Comment (shanbady) [inline src/SetCriteria.js:17]: remove semi-colons

File: src/EditCriteriaForm.js [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    }, {});
+    var keys = [];
+    for (let i = 0; i < data_rows.length; i++) {
+      let spec_id = data_rows[i].criterion_option_spec_id;
+      if (!keys.includes(spec_id)) {
+        keys.push(spec_id);
+      }
+    }
+
+    this.setState({ rows, keys });
```
Comment (shanbady) [inline src/EditCriteriaForm.js:62]: setting the state will cause the component to re-render. im wondering if this is responsible for the infinite loop we are seeing with requests.

**Review by shanbady — CHANGES_REQUESTED** [review]

**Comment by shanbady** [comment]
I am also seeing the behavior that @pbx described. if you open up dev-tools on this interface you will see that it makes a never ending series of requests to the analyze endpoint.

test

File: src/SetCriteria.js (hunk 4 above)
Comment (jonkiparsky) [reply src/SetCriteria.js:17]: Semicolons are required in our standards. 
https://github.com/masschallenge/standards/blob/master/js_standards.md#432-semicolons-are-required

**Review by shanbady — APPROVED** [review]

**Comment by shanbady** [comment]
merging this as-is since I think i can grep the functionality I need to cherry-pick over into the directory repo from this PR.

**Comment by pbx** [comment]
@shanbady  @jonkiparsky  -- I had assigned myself to this, was reviewing it today and returned to it this evening, but since it's now merged I am letting it go. However I'll note that I saw some JS errors in the console on `/app-allocator-setup/` after hitting "Submit" on changes from `/set-criteria/`. The first appeared about five seconds after load:
```
Warning: Can't call setState (or forceUpdate) on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in the componentWillUnmount method.
    in EditCriteriaForm (created by Route)
    in Route (created by withRouter(EditCriteriaForm))
    in withRouter(EditCriteriaForm) (at SetCriteria.js:42)
    in div (at SetCriteria.js:38)
    in SetCriteria (created by Route)
    in Route (at app.js:12)
```
The next appeared a few seconds after that:
```
Warning: Each child in an array or iterator should have a unique "key" prop.

Check the render method of `AnalyzeJudgingRoundDisplay`. See https://fb.me/react-warning-keys for more information.
    in li (at AnalyzeJudgingRoundDisplay.js:33)
    in AnalyzeJudgingRoundDisplay (at ApplicationAllocatorSetup.js:21)
    in div (at ApplicationAllocatorSetup.js:17)
    in ApplicationAllocatorSetup (created by Route)
    in Route (at app.js:11)
    in Switch (at app.js:10)
    in Router (created by BrowserRouter)
    in BrowserRouter (at app.js:9)
    in App (at index.js:8)
```
Are these expected? "Memory leak in your application" is an alarming phrase; if it is to be ignored we should document that. If not, we should fix it. Or if it's truly harmless, find a way to squelch it.

**Comment by VincentHokie** [comment]
`Each child in an array or iterator should have a unique "key" prop.` - this one usually comes up if you're iterating through something and creating a component for some/ all of the iterations without giving the component you're creating a `key` prop. You could use an incrementing count to set this prop/ the ID of the resource that the component is created from (if there is one).

As far as `Can't call setState (or forceUpdate) on an unmounted component....` - I haven't dug around much, but my guess would be it's caused by calling setState in the componentdidupdate lifecycle method. See [this link](https://reactjs.org/docs/react-component.html#componentdidupdate).
